### PR TITLE
Generate waveform polarizations

### PR DESCRIPTION
**Pull request by vaishakp**
Generate waveform polarizations for an observer from NR data.

### CODE REVIEW

File: nrcatalogtools/lvc.py [hunk 1]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
     )
+
+
+
+
+##################################
+# Generate time domain waveforms
+##################################
+
+def Norm(vec_x, vec_y, vec_z):
```
Comment (prayush) [inline nrcatalogtools/lvc.py:235]: We can replace this with `https://numpy.org/doc/stable/reference/generated/numpy.linalg.norm.html`, i.e.

`Norm(x, y, z)` is identical to `numpy.linalg.norm( [x, y, z] )`

File: nrcatalogtools/lvc.py [hunk 2]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
+    vec_x, vec_y, vec_z : float
+                          The components of a three-vector
+
+    Returns
+    -------
+    vec_x, vec_y, vec_z : float
+                          The normalized components of the three-vector.
+    '''
+
+    norm_vec = Norm(vec_x, vec_y, vec_z)
```
Comment (prayush) [inline nrcatalogtools/lvc.py:267]: Use `numpy.linalg.norm([vec_x, vec_y, vec_z])` here.

File: nrcatalogtools/lvc.py [hunk 3]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
+
+    return vec_x/norm_vec, vec_y/norm_vec, vec_z/norm_vec\
```
Comment (prayush) [inline nrcatalogtools/lvc.py:269]: Remove trailing slash "\"

File: nrcatalogtools/lvc.py [hunk 4]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
+
+def CrossProduct(vec1_x, vec1_y, vec1_z, vec2_x, vec2_y, vec2_z):
```
Comment (prayush) [inline nrcatalogtools/lvc.py:271]: Can be replaced by https://numpy.org/doc/stable/reference/generated/numpy.cross.html

ie `CrossProduct(vec1_x, vec1_y, vec1_z, vec2_x, vec2_y, vec2_z)` is the same as `numpy.cross([vec1_x, vec1_y, vec1_z], [vec2_x, vec2_y, vec2_z])`

File: nrcatalogtools/lvc.py [hunk 5]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
+    from scipy.interpolate import interp1d
+    time, freq = H5File.attrs['Omega-vs-time']
+    
+    
+    
+    RefTime = interp1d(freq, time, kind='cubic')[ref_freq]
+    
+    return RefTime
+
+def CheckNRAttrs(H5File, ReqAttrs):
```
Comment (prayush) [inline nrcatalogtools/lvc.py:365]: Would it be useful to have a reasonable default for `ReqAttrs` here given the LVCNR format is, per-se, fixed?
For e.g. one could gather all the attributes being read in from H5 files here in `lvc.py` and collect them as a list here..

File: nrcatalogtools/lvc.py [hunk 6]
```diff
@@ -224,3 +224,462 @@ def get_strain_from_lvcnr_file(path_to_file,
...
+    ReqTSAttrs : list
+               A list of attribute keys.
+    ref_time : float
+            Reference time.
+    Returns
+    -------
+    params : dict
+             The parameter values at the reference time.
+
+    Notes
```
Comment (prayush) [inline nrcatalogtools/lvc.py:412]: Perhaps remove this empty section "Notes"

**Comment by prayush** [comment]
@vaishakp this PR is quickly becoming quite large. It might be easier to factor out smaller pieces of it and get them merged first and faster.

File: nrcatalogtools/legacy.py [hunk 7]
```diff
@@ -0,0 +1,477 @@
+import os
```
Comment (adivijaykumar) [inline nrcatalogtools/legacy.py:1]: This file was retired in #10, please remove it.

File: nrcatalogtools/lvc.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 import h5py
 import lal
 import lalsimulation as lalsim
+import numpy as np
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:4]: duplicate import, remove

File: nrcatalogtools/lvc.py [hunk 9]
```diff
@@ -705,78 +716,84 @@ def get_nr_to_lal_rotation_angles(
...
             # No interpolation required. Use available reference values.
-            Interp = False
+            interp = False
 
         else:
             # Try to get the reference time from orbital frequency
             try:
-                TRef = GetRefTimeFromRefFreq(H5File, FRef)
+                t_ref = get_ref_time_from_ref_freq(h5_file, f_ref)
+
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:732]: Whitespace issue. Use `black` to format files.

File: nrcatalogtools/lvc.py [hunk 10]
```diff
@@ -705,78 +716,84 @@ def get_nr_to_lal_rotation_angles(
...
-                RefParams = GetRefVals(Metadata)
+                # LAL source frame is present in the simulation metadata
+                ref_params = get_ref_vals(sim_metadata)
         else:
-            # print(RefCheckDefH5, AbsentAttrsH5)
-            RefParams = GetRefVals(H5File)
-            # print(RefParams)
-    elif Interp is True:
+            ref_params = get_ref_vals(h5_file)
+
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:789]: Whitespace issue. Use `black` to format files.

File: nrcatalogtools/lvc.py [hunk 11]
```diff
@@ -705,78 +716,84 @@ def get_nr_to_lal_rotation_angles(
...
             # Try to get the reference time from orbital frequency
             try:
-                TRef = GetRefTimeFromRefFreq(H5File, FRef)
+                t_ref = get_ref_time_from_ref_freq(h5_file, f_ref)
+                
                 # Check if interpolation is required
-                Interp, avail_ref_time = CheckInterpReq(H5File, TRef)
-            except Exception:
+                interp, avail_ref_time = check_interp_req(h5_file, t_ref)
+            except:
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:735]:
```suggestion
            except Exception:
```

File: nrcatalogtools/lvc.py [hunk 12]
```diff
@@ -789,41 +806,42 @@ def get_nr_to_lal_rotation_angles(
...
+        ref_check_interp_req, absent_interp_attrs = check_nr_attrs(
+            h5_file, req_ts_attrs
+        )
 
-        if RefCheckInterpReq is False:
+        if ref_check_interp_req is False:
             raise Exception(
-                "Insufficient information to compute the LAL source frame at given reference time.",
-                f"Missing information is {AbsentInterpAttrs}.",
+                f"Insufficient information to compute the LAL source frame at given reference time. Missing information is {absent_interp_attrs}."
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:815]: Ensure line length is less than 120 by breaking the string up.

File: nrcatalogtools/lvc.py [hunk 13]
```diff
@@ -705,78 +716,84 @@ def get_nr_to_lal_rotation_angles(
...
+                )
+
+                if ref_check_def_meta_sxs is True:
+                    # Compute the LAL source frame from simulation metadata
+                    ref_params = compute_lal_source_frame_from_sxs_metadata(
+                        sim_metadata
+                    )
+                else:
+                    raise Exception(
+                        f"Insufficient information to compute the LAL source frame.\n Missing information is {absent_attrs_meta_sxs}."
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:782]: Ensure line length is less than 120 by breaking the string up.

**Review by adivijaykumar — CHANGES_REQUESTED** [review]

File: nrcatalogtools/lvc.py (hunk 12 above)
Comment (prayush) [reply nrcatalogtools/lvc.py]: I suggest keeping the original format

```
                "Insufficient information to compute the LAL source frame at given reference time.",
                f"Missing information is {absent_interp_attrs}."
```

File: nrcatalogtools/lvc.py (hunk 13 above)
Comment (prayush) [reply nrcatalogtools/lvc.py]:
```
                        "Insufficient information to compute the LAL source frame.\n"
                        f" Missing information is {absent_attrs_meta_sxs}."
```

File: nrcatalogtools/maya.py [hunk 14]
```diff
@@ -31,8 +31,6 @@ def __init__(self, catalog=None, use_cache=True, verbosity=0, **kwargs) -> None:
         self.waveform_data_url = utils.maya_catalog_info["data_url"]
         self.waveform_data_dir = utils.maya_catalog_info["data_dir"]
 
-        self._add_paths_to_metadata()
```
Comment (prayush) [inline nrcatalogtools/maya.py:34]: Why are you deleting this call? It is needed to keep metadata/waveform data filepaths available in the catalog object's attributes.

File: nrcatalogtools/waveform.py [hunk 15]
```diff
@@ -1,9 +1,10 @@
-import lal
-import numpy as np
 import os
+
 import h5py
+import lal
+import numpy as np
 from sxs import WaveformModes as sxs_WaveformModes
-
+from nrcatalogtools.lvc import GetNRToLALRotationAngles
```
Comment (prayush) [inline nrcatalogtools/waveform.py:7]: Change to

```
from nrcatalogtools.lvc import get_nr_to_lal_rotation_angles
```

File: nrcatalogtools/waveform.py [hunk 16]
```diff
@@ -232,12 +262,63 @@ def get_td_waveform(
...
+        # Note: 02 May 23 (VP)
+        # Presently, coa_phase is not implemented.
+        print(
+            "Warining! coa_phase is not implemented yet. The reference phase will be used."
+        )
+
+        # Compute angles
+        with h5py.File(self.filepath) as h5_file:
+            # print(H5File.attrs.keys())
+            angles = GetNRToLALRotationAngles(
```
Comment (prayush) [inline nrcatalogtools/waveform.py:311]: Change to
```
angles = get_nr_to_lal_rotation_angles(
```

File: nrcatalogtools/lvc.py (hunk 12 above)
Comment (prayush) [reply nrcatalogtools/lvc.py]: nvm, these have alreayd been fixed

**Comment by adivijaykumar** [comment]
@vaishakp how much time do you estimate PR is going to take before it becomes ready? The current `master` has breaking changes that would be resolved only by this PR (see #22 for a PR that will needs #3 to be merged). If it will take more time, I would suggest that we split this into separate PRs, with one of them only having basic formatting fixes that were breaking `flake8`.

**Comment by vaishakp** [comment]
I will push the changes in an hour.

**Comment by vaishakp** [comment]
This can now be merged. Testing is not yet completed. I will open a new PR for tests.

File: nrcatalogtools/waveform.py [hunk 17]
```diff
@@ -1,9 +1,10 @@
-import lal
-import numpy as np
 import os
+
 import h5py
+import lal
+import numpy as np
 from sxs import WaveformModes as sxs_WaveformModes
-
+from nrcatalogtools.lvc import compute_phi_ref, get_nr_to_lal_rotation_angles
```
Comment (adivijaykumar) [inline nrcatalogtools/waveform.py:7]:
```suggestion
from .lvc import compute_phi_ref, get_nr_to_lal_rotation_angles
```

File: nrcatalogtools/waveform.py [hunk 18]
```diff
@@ -232,12 +265,60 @@ def get_td_waveform(
...
+
+        Parameters
+        ----------
+        inclination : float
+                      The inclination angle of the observer
+                      in the LAL source frame
+        coa_phase : float
+                    The coalescence phase. This will be
+                    used to compute the reference orbital phase.
+        fref, tref : float, optional
```
Comment (adivijaykumar) [inline nrcatalogtools/waveform.py:292]:
```suggestion
        f_ref, t_ref : float, optional
```

File: nrcatalogtools/lvc.py [hunk 19]
```diff
@@ -302,60 +332,62 @@ def get_ref_time_from_ref_freq(H5File, ref_freq):
...
+        sim_metadata_object : h5 file object, dict
+                         The NR h5py file handle or simulation metadata.
+        req_attrs : list
+                   A list of attribute keys.
+        Returns
+        -------
+        present : bool
+                  Whether or not all specified attributes are present.
+    h = self.interpo    absent_attrs : list
+                     The attributes that are absent.
```
Comment (adivijaykumar) [inline nrcatalogtools/lvc.py:360]: There is something wrong here. Why is there a `h=self.interpo`?

**Review by adivijaykumar — CHANGES_REQUESTED** [review]
Leaving some comments about formatting issues.

File: nrcatalogtools/lvc.py [hunk 20]
```diff
@@ -623,26 +659,83 @@ def transform_spins_nr_to_lal(nrSpin1, nrSpin2, n_hat, ln_hat):
...
+    -------
+    obs_phi_ref : float
+                  The orbital reference phase for the observer at NR
+                  reference time.
+    """
+    if (t_ref is not None) or (f_ref is not None):
+        # Compute t_ref from f_ref
+        raise NotImplementedError("f_ref input is presently not supported!")
+
+        # ToDo : Use waveformtools.differentiate here (04.05.23, VP)
```
[the remark this answers is not in the document]
Comment (adivijaykumar) [reply nrcatalogtools/lvc.py:695]: IMO this is exactly what github issues are for

File: nrcatalogtools/waveform.py (hunk 15 above)
Comment (vaishakp) [reply nrcatalogtools/waveform.py]: In my opinion, relative imports can bring in inconsistencies, especially for us.

Comment (prayush) [reply nrcatalogtools/waveform.py]: Agree with @vaishakp. Lets avoid relative imports as much as possible.

File: nrcatalogtools/lvc.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (prayush) [reply nrcatalogtools/lvc.py:695]: Sure, we can open an issue also. Lets leave the commit in anyway.

File: nrcatalogtools/lvc.py (hunk 19 above)
Comment (prayush) [reply nrcatalogtools/lvc.py:360]: seems like a copy-pasta error

File: nrcatalogtools/waveform.py (hunk 15 above)
Comment (adivijaykumar) [reply nrcatalogtools/waveform.py]: OK, I am fine with that. @vaishakp however could you please elaborate what inconsistencies you are talking about?

File: nrcatalogtools/lvc.py (hunk 19 above)
Comment (vaishakp) [reply nrcatalogtools/lvc.py]: True. I'm having issues with mouse!

**Pull request merged**